### PR TITLE
Add runtime metrics reporting and set it to start by default

### PR DESCRIPTION
**Pull request by madhuravi**
Add runtime metrics reporting and set it to start by default

### CODE REVIEW

File: metrics/runtime_metrics.go [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+func StartReportingRuntimeMetrics(
+	scope tally.Scope, reportInterval time.Duration,
+) *RuntimeReporter {
+	runtimeReporter := NewRuntimeReporter(scope, reportInterval)
+	runtimeReporter.Start()
+	return runtimeReporter
+}
+
+type runtimeMetrics struct {
+	NumGoRoutines   tally.Gauge
```
Comment (sectioneight) [inline metrics/runtime_metrics.go]: Why are most of these exported except for the `lastNumGC`? It seems like, since this struct isn't exported, these should all be unexported fields as well.

File: metrics/runtime_metrics.go [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+}
+
+// IsRunning returns true if the reporter has been started and false if not.
+func (r *RuntimeReporter) IsRunning() bool {
+	return r.started
+}
+
+// Start starts the reporter thread that periodically emits metrics.
+func (r *RuntimeReporter) Start() {
+	if r.started {
```
Comment (sectioneight) [inline metrics/runtime_metrics.go]: Intentionally not thread-safe? Should we leave a note about this?

File: metrics/runtime_metrics_test.go [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+func TestStartRuntimeReporter(t *testing.T) {
+	reporter := StartReportingRuntimeMetrics(tally.NoopScope, time.Millisecond)
+	assert.True(t, reporter.IsRunning())
+
+	runtime.GC()
+	time.Sleep(5 * time.Millisecond)
+	reporter.report()
+	time.Sleep(5 * time.Millisecond)
+	reporter.report()
+	reporter.Close()
```
Comment (sectioneight) [inline metrics/runtime_metrics_test.go]: What is this actually asserting? And what's with the magic `5` milliseconds?

File: modules/uhttp/client/client.go [hunk 4]
```diff
@@ -43,6 +43,9 @@ func New(client *http.Client, filters ...Filter) *Client {
 	return &Client{Client: client, filters: filters}
 }
 
+// DefaultClient is the default uhttp client with defaults for http client and filters.
+var DefaultClient = New(http.DefaultClient)
```
Comment (sectioneight) [inline modules/uhttp/client/client.go]: This seems unrelated. Did you accidentally add this to this PR?

File: modules/uhttp/filters.go [hunk 5]
```diff
@@ -55,7 +55,7 @@ func tracingServerFilter(ctx fx.Context, w http.ResponseWriter, r *http.Request,
 	ext.HTTPUrl.Set(span, r.URL.String())
 	defer span.Finish()
 
-	gctx := fx.NewContext(opentracing.ContextWithSpan(ctx, span), ctx)
+	gctx := ctx.WithContext(opentracing.ContextWithSpan(ctx, span))
```
Comment (sectioneight) [inline modules/uhttp/filters.go]: again, unrelated?

File: service/core.go [hunk 6]
```diff
@@ -64,19 +66,20 @@ type SetContainerer interface {
...
+	scopeMux        sync.Mutex
+	scope           tally.RootScope
+	runtimeReporter *metrics.RuntimeReporter
+	observer        Observer
+	resources       map[string]interface{}
+	logConfig       ulog.Configuration
+	log             ulog.Log
+	tracerConfig    jaegerconfig.Configuration
+	tracer          opentracing.Tracer
+	tracerCloser    io.Closer
```
Comment (sectioneight) [inline service/core.go]: if we're going to be messing up history here can we alphabetize these? this is getting pretty big and hard to read

File: service/service.go [hunk 7]
```diff
@@ -139,6 +141,10 @@ func New(options ...Option) (Owner, error) {
 		metrics.Freeze()
 	}
 
+	if svc.RuntimeMetrics() == nil {
```
Comment (sectioneight) [inline service/service.go]: We should provide a way to opt-out. Somebody should be able to explicitly say "I don't want to report runtime metrics" (as an option).

**Comment by madhuravi** [comment]
@yurishkuro since you wrote the original code.

**Comment by madhuravi** [comment]
@uber-go/service-framework

File: metrics/runtime_metrics.go (hunk 1 above)
Comment (madhuravi) [inline metrics/runtime_metrics.go]: You're right, changed all to unexported.

File: service/core.go (hunk 6 above)
Comment (madhuravi) [inline service/core.go]: Alphabetized and grouped some fields into embedded structs.

File: modules/uhttp/filters.go (hunk 5 above)
Comment (madhuravi) [inline modules/uhttp/filters.go]: removed.

File: modules/uhttp/client/client.go (hunk 4 above)
Comment (madhuravi) [inline modules/uhttp/client/client.go]: removed.

File: metrics/runtime_metrics_test.go (hunk 3 above)
Comment (madhuravi) [inline metrics/runtime_metrics_test.go]: This was copied over but nothing magic about 5 afaik. Runtime metrics are reported every millsecond, even 1 millisecond should do it.

File: metrics/runtime_metrics.go [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+import (
+	"runtime"
+	"sync/atomic"
+	"time"
+
+	"github.com/uber-go/tally"
+)
+
+// StartReportingRuntimeMetrics starts reporting runtime metrics under given metrics scope with
+// given frequency.
```
Comment (glibsm) [inline metrics/runtime_metrics.go]: this wrapping is all out of whack. Let stick to 80 char doc strings, if we can

File: service/core.go [hunk 9]
```diff
@@ -63,20 +65,49 @@ type SetContainerer interface {
...
+	tracer       opentracing.Tracer
+	tracerCloser io.Closer
+	tracerConfig jaegerconfig.Configuration
+}
+
+func (tc *tracerCore) Tracer() opentracing.Tracer {
+	return tc.tracer
+}
+
+type loggingCore struct {
```
Comment (glibsm) [inline service/core.go]: what's all this for?

File: metrics/runtime_metrics.go [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+	memoryHeapIdle  tally.Gauge
+	memoryHeapInuse tally.Gauge
+	memoryStack     tally.Gauge
+	numGC           tally.Counter
+	gcPauseMs       tally.Timer
+	lastNumGC       uint32
+}
+
+// RuntimeReporter is a struct containing the state of the RuntimeMetricsReporter.
+type RuntimeReporter struct {
```
Comment (glibsm) [inline metrics/runtime_metrics.go]: this is going to get into hand-waving territory, but calling it a Reporter is very confusing, since it takes in a tally.Scope, which has a Reporter to send metrics over.

I'd prefer we call it something along the lines of "runtime metrics collector", since that's what it actually does.

File: service/service.go (hunk 7 above)
Comment (madhuravi) [inline service/service.go]: Good point. Why do we want this only for runtime metrics? We don't seem to have it for standard metrics. It is running in a subroutine, should not cause a latency impact.

File: metrics/runtime_metrics.go (hunk 8 above)
Comment (madhuravi) [inline metrics/runtime_metrics.go]: I thought we decided on 100 char?

File: service/core.go (hunk 9 above)
Comment (madhuravi) [inline service/core.go]: Just grouping to make it more readable/manageable. Feedback is welcome.

File: metrics/runtime_metrics.go (hunk 10 above)
Comment (madhuravi) [inline metrics/runtime_metrics.go]: It does actually report though. It delegates to the tally reporter but that's an implementation detail.

Comment (glibsm) [inline metrics/runtime_metrics.go]: I don't follow. Where does the delegation happen?

Comment (madhuravi) [inline metrics/runtime_metrics.go]: never mind, terminology mixup. Will change it.

**Comment by sectioneight** [comment]
Somebody could pass in a noop metrics scope which would be non nil. But
there's no way for somebody to disable this behavior as is.

**Comment by madhuravi** [comment]
@sectioneight if they pass in a noop metrics scope, runtime will not get reported either since it uses the same root scope. It does still collect metrics though. I can create a NoopRuntimeCollector that noops on all methods but I feel like we can punt on that until we have real use cases where they would want to disable this altogether. What do you think?

**Comment by sectioneight** [comment]
It's still doing work that could be undesired from a service. We can make
an Option that disables it, no need to create a stub implementation. Just
check if the Option was passed or if a config field is set to explicitly
disable it.

File: metrics/runtime_metrics.go (hunk 2 above)
Comment (madhuravi) [inline metrics/runtime_metrics.go]: Changed my mind, makes sense to add one here. Done.

File: metrics/runtime_metrics.go [hunk 11]
```diff
@@ -0,0 +1,168 @@
...
+	"sync/atomic"
+	"time"
+
+	"github.com/uber-go/tally"
+)
+
+// StartCollectingRuntimeMetrics starts generating runtime metrics under given metrics scope with
+// given frequency.
+// Recommended usage:
+//     rmr := StartCollectingRuntimeMetrics(rootScope.Scope("runtime"), time.Second)
```
Comment (anuptalwalkar) [inline metrics/runtime_metrics.go]: This would be a nice addition in readme, doc.go

File: metrics/runtime_metrics.go [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+	r.metrics.memoryStack.Update(int64(memStats.StackInuse))
+
+	// memStats.NumGC is a perpetually incrementing counter (unless it wraps at 2^32)
+	num := memStats.NumGC
+	lastNum := atomic.SwapUint32(&r.metrics.lastNumGC, num) // reset for the next iteration
+	if delta := num - lastNum; delta > 0 {
+		r.metrics.numGC.Inc(int64(delta))
+		if delta > 255 {
+			// too many GCs happened, the timestamps buffer got wrapped around. Generate only the last 256
+			lastNum = num - 256
```
Comment (anuptalwalkar) [inline metrics/runtime_metrics.go]: why 256? make it a package variable?

File: service/core.go (hunk 9 above)
Comment (anuptalwalkar) [inline service/core.go]: We can go a little further and move logging initialization code here or under /service/core package along with other *Core stuff, and use function call in the service New. service.New is about 100+ lines of code right now.

File: service/service.go [hunk 13]
```diff
@@ -139,6 +141,17 @@ func New(options ...Option) (Owner, error) {
 		metrics.Freeze()
 	}
 
+	if svc.RuntimeMetricsCollector() == nil {
+		var runtimeMetricsConfig metrics.RuntimeConfig
+		err := svc.configProvider.Get("runtimeMetrics").PopulateStruct(&runtimeMetricsConfig)
```
Comment (glibsm) [inline service/service.go]: we already have `metrics` top-level struct for config. Perhaps we should nest `runtime` under that?

File: service/core.go (hunk 9 above)
Comment (glibsm) [inline service/core.go]: I'm not sure we should do this refactor as part of this diff, it would be much easier to iterate on and discuss as part of a separate diff

File: metrics/runtime_metrics_test.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+func TestRuntimeCollector(t *testing.T) {
+	collector := NewRuntimeCollector(tally.NoopScope, time.Millisecond)
+	defer closeCollector(t, collector)
+
+	assert.False(t, collector.IsRunning())
+	collector.Start()
+	assert.True(t, collector.IsRunning())
+	runtime.GC()
+	time.Sleep(5 * time.Millisecond)
+	collector.generate()
```
Comment (glibsm) [inline metrics/runtime_metrics_test.go]: after collector.generate, can we validate anything?

File: metrics/runtime_metrics.go (hunk 11 above)
Comment (madhuravi) [inline metrics/runtime_metrics.go]: Good callout. Looks like metrics doesn't have docs yet. Will add this as part of a separate diff.

File: service/service.go (hunk 13 above)
Comment (madhuravi) [inline service/service.go]: Who manages top level metrics? I don't want to accidentally override a preconfigured key for eg.

File: metrics/runtime_metrics_test.go (hunk 14 above)
Comment (glibsm) [inline metrics/runtime_metrics_test.go]: No need for mocks here either.

Just right above in this test, we are using `tally.Noopscope`. Instead of that we should have a `memReporter` that just creates a hash of reported metrics (I think we should already have that somewhere, and if not that's pretty easy to build and we can rebase onto that diff).

Then you don't have to mock anything, you can assert on that reporter directly to make sure it has the correct metrics.

File: service/core.go (hunk 9 above)
Comment (madhuravi) [inline service/core.go]: This was originally from Aiden's comment about the struct getting bloated. Since it was a result of adding a new field to this struct, I think it's ok to have it as part of this diff. 
Good idea Anup, don't want to add all that refactor as part of this though. Will make it part of a new diff.

File: service/service.go (hunk 13 above)
Comment (sectioneight) [inline service/service.go]: We manage it, we can make it whatever we want. I agree with Glib, this should be nested `metrics.runtime`

Comment (madhuravi) [inline service/service.go]: Done

File: metrics/runtime_metrics_test.go (hunk 14 above)
Comment (madhuravi) [inline metrics/runtime_metrics_test.go]: Based on offline discussion with @sectioneight @glibsm @anuptalwalkar the conclusion is we only test our library's functionality. The memStatsReporter is for testing that tally has recorded these metrics (outside uberfx functionality), so decided to not do this.

File: metrics/runtime_metrics.go [hunk 15]
```diff
@@ -0,0 +1,171 @@
...
+			lastNum = num - _numGCThreshold
+		}
+		for i := lastNum; i != num; i++ {
+			pause := memStats.PauseNs[i%256]
+			r.metrics.gcPauseMs.Record(time.Duration(pause))
+		}
+	}
+}
+
+// Close stops collecting runtime metrics. It cannot be started again after it's
```
Comment (glibsm) [inline metrics/runtime_metrics.go]: @anuptalwalkar maybe we should consider this strategy for service start/stop... might make handling states and cleanup a lot easier. If a service is stopped - that's it. Re-create it

File: metrics/runtime_metrics_test.go [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+func TestStartRuntimeCollector(t *testing.T) {
+	collector := StartCollectingRuntimeMetrics(tally.NoopScope, time.Millisecond, _emptyRuntimeConfig)
+	defer closeCollector(t, collector)
+
+	assert.True(t, collector.IsRunning())
+	runtime.GC()
+	time.Sleep(time.Millisecond)
+	collector.generate()
```
Comment (glibsm) [inline metrics/runtime_metrics_test.go]: Are these generate calls actually necessary? it looks like the collector works on a ticker, which means every time we sleep for `+-time.Millisecond` it will run on it's own (runtime_metrics.go:126).

I think the output of this test is that the collection has been done four times. If only there was a way to find out how many metrics ended up in tally ;)

File: service/service_test.go [hunk 17]
```diff
@@ -121,6 +121,22 @@ func TestServiceWithRoles(t *testing.T) {
...
+metrics:
+  runtime:
+    disabled: true
+`)
+	cfgOpt := WithConfiguration(config.NewYAMLProviderFromBytes(data))
+
+	svc, err := New(cfgOpt)
+	require.NoError(t, err)
+
+	assert.Nil(t, svc.RuntimeMetricsCollector())
```
Comment (glibsm) [inline service/service_test.go]: nice!

File: metrics/runtime_metrics_test.go (hunk 16 above)
Comment (madhuravi) [inline metrics/runtime_metrics_test.go]: Good point. Removed.
